README.md
# Programmatic Music Generation

Programmatically generate music using the Sonic-Pi framework.

## Conceptual Layout of Song

A song is outlined in various conceptual units that are able to be programmatically generated.

### Mood

### Structure

### Chords

### Song Layers


## Control Surfaces

Each piece will have the design choices able to be selected by interfaceing with User controlled entities. These can be MIDI or electro-mechanical devices of any kind that are tied to values within the Sonic-Pi implementation.

In short, if there is a choice to be made, then that choice or range of opitions is mapped to a variable in Sonic-Pi. 
The selection of that choice is then subsequently tied to a real world selector, be it MIDI channel, OSC interface, analog hardware input, etc.
This enables the input choices made about the song layout able to be controlled via a range of user inputs...

CHANGELOG.md
# changelog for `prog_music`

## [Unreleased]

### Changed

### Added

### Fixed

### Removed

### Security

---

## [0.1.0] - 2021-

### Changed

### Added in v0.1.0

- `CHANGELOG.md`: A changelog file following SemVer is added.
- `README.md`: A readme is added.

### Fixed

### Removed

### Security

sonic-pi/tilburg2.rb
# Tilburg 2

# Coded by Sam Aaron

use_debug false
load_samples :guit_em9, :bd_haus

live_loop :low do
  tick
  synth :zawa, wave: 1, phase: 0.25, release: 5, note: (knit :e1, 12, :c1, 4).look, cutoff: (line 60, 120, steps: 6).look
  sleep 4
end

with_fx :reverb, room: 1 do
  live_loop :lands, auto_cue: false do
    use_synth :dsaw
    use_random_seed 310005
    ns = (scale :e2, :minor_pentatonic, num_octaves: 4).take(4)
    16.times do
      play ns.choose, detune: 12, release: 0.1, amp: 2, amp: rand + 0.5, cutoff: rrand(70, 120), amp: 2
      sleep 0.125
    end
  end
end

live_loop :fietsen do
  sleep 0.25
  sample :guit_em9, rate: -1
  sleep 7.75
end

# live_loop :tijd do
#   sample :bd_haus, amp: 2.5, cutoff: 100
#   sleep 0.5
# end

# live_loop :ind do
#   sample :loop_industrial, beat_stretch: 1
#   sleep 1
# end

 

python/python-tilburg.py
from psonic import *
from threading import Thread, Condition
import random


def random_riff(condition):
    use_synth(PROPHET)
    sc = scale(E3, MINOR)
    while True:
        s = random.choice([0.125, 0.25, 0.5])
        with condition:
            condition.wait()  # Wait for message
        for i in range(8):
            r = random.choice([0.125, 0.25, 1, 2])
            n = random.choice(sc)
            co = random.randint(30, 100)
            play(n, release=r, cutoff=co)
            sleep(s)


def drums(condition):
    while True:
        with condition:
            condition.notifyAll()  # Message to threads
        for i in range(16):
            r = random.randrange(1, 10)
            sample(DRUM_BASS_HARD, rate=r)
            sleep(0.125)


condition = Condition()
random_riff_thread = Thread(name="consumer1", target=random_riff, args=(condition,))
drums_thread = Thread(name="producer", target=drums, args=(condition,))

random_riff_thread.start()
drums_thread.start()

input("Press Enter to continue...")

sonic-pi/toy_dinosaur_in_a_finnish_sauna.rb
# Test-driving the new Mehackit
# samples by making some beats

# First, we begin the story with our
# little TOY DINO in a Finnish sauna!

use_bpm 122

with_fx :reverb, room: 0.3, mix: 0.7 do
  sample :mehackit_robot6, rate: 0.5, amp: 1.2, start: 0.4, attack: 0.2
  sleep 2
  sample :mehackit_robot1, start: 0.0, attack: 0.25, cutoff: 120, rate: 0.75, amp: 2
  sleep 1
  sample :mehackit_robot3, cutoff: 120
  sleep 4
  sample :mehackit_robot2, cutoff: 120
  sleep 2
end

with_fx :reverb, room: 0.9 do
  sample :ambi_sauna, amp: 0.85
  sleep 4
  
  with_fx :reverb, room: 0.2, mix: 0.65 do
    sample :mehackit_robot5, rate: -0.5, amp: 2
    sleep 6
  end
end

4.times do |i|
  
  in_thread do
    8.times do
      if i >= 3
        hipass = 70
      else
        hipass = 0
      end
      sample :bd_mehackit, amp: 1.5, hpf: hipass
      sleep 1
    end
  end
  
  with_fx :reverb, room: 0.4, mix: 0.6 do
    in_thread do
      with_fx :lpf, res: 0.7, cutoff: 130, cutoff_slide: 8 do |c|
        if i == 3
          control c, cutoff: 50
        end
        with_fx :distortion, amp: 0.7 do
          sample :glitch_bass_g, rate: -1, amp: 1.9
        end
        2.times do
          sleep 3
          sample [:glitch_perc1, :glitch_perc2, :glitch_perc3, :glitch_perc4, :glitch_perc5].choose, rate: rrand(0.8, 1.2)*[1,1,-1].choose
          sleep 0.5
          sample [:glitch_perc1, :glitch_perc2, :glitch_perc3, :glitch_perc4, :glitch_perc5].choose, rate: rrand(0.8, 1.2)*[1,1,-1].choose
          sleep 0.5
        end
      end
    end
  end
  
  in_thread do
    with_fx :lpf, res: 0.7, cutoff: 130, cutoff_slide: 8 do |c|
      if i == 3
        control c, cutoff: 50
      end
      with_fx :panslicer, wave: 2, phase: 2.5, mix: 0.4 do
        32.times do
          sample [:loop_mehackit1, :loop_mehackit2].choose, beat_stretch: 4, rate: [1,1,1,1,-1,1.5].choose, amp: 1, slice: pick, sustain: 0, release: 0.2
          sleep 0.25
        end
      end
    end
  end
  
  in_thread do
    with_fx :lpf, res: 0.7, cutoff: 130, cutoff_slide: 8 do |c|
      if i == 3
        control c, cutoff: 50
      end
      2.times do
        sample [:mehackit_phone1, :mehackit_phone2, :mehackit_phone3, :mehackit_phone4, :mehackit_phone5].choose, rate: rrand(0.8, 1.2)*[1,1,-1].choose
        sleep 1.5
      end
      sleep 1.0
      2.times do
        sample [:mehackit_robot1, :mehackit_robot2, :mehackit_robot3, :mehackit_robot4].choose, rate: rrand(0.8, 1.2)*[1,1,-1].choose
        sleep 1.5
      end
      sleep 0.5
    end
  end
  
  if i == 3
    with_fx :reverb, room: 0.99 do
      sample :loop_3d_printer, amp: 2.6, rate: -1, attack: 8, beat_stretch: 8, hpf: 50
    end
  end
  sleep 8
  sample :perc_impact1
end

sleep 1
with_fx :reverb, room: 0.99 do
  sample :perc_bell2, hpf: 60
  sample :perc_impact1, rate: -0.7, amp: 2
end
sleep 3

4.times do |i|
  
  with_fx :reverb, room: 0.65 do
    with_fx :distortion, distort: 0.9, amp: 0.325 do
      in_thread do
        if i < 2
          bassroot = :G1
        else
          bassroot = :A1
        end
        h = synth :chipbass, note: bassroot, note_slide: 16, cutoff: 60, cutoff_slide: 4, attack: 2, sustain: 4, release: 0.1
        sleep 1
        control h, note: bassroot+48, cutoff: 130
        control h, cutoff: 130
        sleep 5
        2.times do
          s = synth :pulse, note: bassroot, note_slide: 2, cutoff: 60, cutoff_slide: 2, sustain: 1, release: 0.1
          control s, note: bassroot+60, cutoff: 130
          sleep 1
        end
      end
    end
  end
  
  in_thread do
    8.times do
      if i >= 3
        hipass = 70
      else
        hipass = 0
      end
      sample :bd_mehackit, amp: 1.5, hpf: hipass
      sleep 1
    end
  end
  
  in_thread do
    with_fx :reverb do
      4.times do
        sleep 1
        sample :sn_generic
        sample :perc_swoosh, amp: 1.5
        sleep 1
      end
    end
  end
  
  in_thread do
    with_fx :bitcrusher, mix: 0.5, bits: 16, bits_slide: 8 do |z|
      control z, bits: 1 if i == 3
      sample :loop_weirdo, beat_stretch: 8, amp: [0, 0.35, 0.7, 1.1][i]
      sleep 8
    end
  end
  
  
  in_thread do
    with_fx :lpf, res: 0.7, cutoff: 130, cutoff_slide: 8 do |x|
      control x, cutoff: 20 if i == 3
      2.times do
        sample :loop_electric, beat_stretch: 4, amp: [1, 1.1, 1.2, 1.3][i]
        sleep 4
      end
    end
  end
  
  if i == 3
    with_fx :reverb, room: 0.99 do
      sample :loop_electric, amp: 3, rate: -1, attack: 8, beat_stretch: 8, hpf: 50
    end
  end
  sleep 8
end

#YEAH!!!

sonic-pi/jupiter.rb
# Alexandre rANGEL
# 'jupiter coming with the wolves' (v13)
# www.quasecinema.org
# 28-Feb-2016 / Sonic Pi 2.12

arrangement = true
mybpmbase = 120
mybpm = mybpmbase # * 2 #6, 8, 12, 16, 4, 2
use_bpm mybpm
startClock = 0 # 0 to start song at beginning
clock = startClock # global var
bar = 0 # global var
set_sched_ahead_time! 1
set_volume! 1.0 #1.1 #0.75
t = Time.new
x = ((t.year - t.month - t.day - t.hour - t.min - t.sec) * 1).to_int
puts "x = #{x}"
use_random_seed = x #1946 1966 1858
# http://www.freesound.org/people/clarinet_pablo_proj/sounds/248870/
# http://www.freesound.org/people/bcnlab/sounds/270736/
# download 2 samples above, rename to rhodes120bpm.wav and clarinetf6.wav
sampleRhodes120bpm =  '/Users/piatra/Documents/SamplesPi/rhodes120bpm.wav' # your path here
sampleClarinetf6 = '/Users/piatra/Documents/SamplesPi/clarinetf6.wav'
load_samples [sampleRhodes120bpm, sampleClarinetf6, :bd_haus, :bd_pure, :bd_klub, :bd_fat]

live_loop :metro do
  
  mybpm = mybpmbase * [0.5,1,1,1,2].choose #ramdom bpms
  mybpm = mybpmbase / 2.0 if bar < 3
  
  with_bpm mybpm do
    cue :metro
    sleep 0.5
    clock = startClock + tick
    bar = (clock / 4) + 1
    puts "let's go!" if clock < 1
    puts "bar : #{bar}, bpm: #{mybpm}"
    puts ring( "1 !","2 ! !","3 ! ! !","4 ! ! ! !")[clock]
  end #mybpm
  
end #metro

with_fx :compressor, slope_below: 2.0, slope_above: 1.15, mix: 0.5 do
  live_loop :song1 do
    
    with_fx :slicer, phase: 4 do
      
      with_bpm mybpm do
        
        # the break
        if bar < (8+16) or bar > (24+16) then
          myAmp = 1.0
        else
          myAmp = 0.0
        end
        
        sample sampleRhodes120bpm, rate: 1.0 / 4.0,
          amp: 12 * myAmp,
          window_size: (ring 0.003333,0.002,0.001)[tick + 1]
        
      end
      
    end #slicer
    
    sleep ( (sample_duration sampleRhodes120bpm) * 4)
    
  end #:song1
  
  sleep 2 # offset
  
  live_loop :song2 do
    if bar > (24+16) then # the break
      
      with_fx :slicer, phase: 4, slope_down: 1.0/8.0 do
        
        with_bpm mybpm do
          
          sample sampleRhodes120bpm, rate: 1.0 / 4.0,
            amp: 7.777 * 1.3,
            window_size: (ring 0.003333,0.002,0.001)[tick],
            pitch: [12,14,16].choose
          
        end #bpm
        
      end #slicer
      
    end #if
    
    sleep ( (sample_duration sampleRhodes120bpm) * 4)
    
  end #:song2
  
  sleep 4
  
  with_fx :echo do
    with_fx :ixi_techno do
      
      live_loop :song3, phase: 0.5 do
        
        # the break
        if bar < (8+16) or bar > (24+16) then
          myAmp = 1.0
        else
          myAmp = 0.0
        end
        
        with_fx :slicer, phase: 8, slope_down: 1.0/8.0 do
          
          with_bpm mybpm do
            
            with_fx :flanger, phase: [0.250,0.5].choose,
            delay: [1,2,4].choose do
              
              with_fx :slicer,
                phase: 0.25 *0.5,
              slope_up: (ring 0,0.1,0,0.25)[tick/1] do
                
                sample sampleRhodes120bpm, rate: 1.0 / 4.0,
                  amp: 5.555 * myAmp,
                  window_size: (ring 0.001,0.003333,0.002,)[tick],
                  pitch: -24
                
              end #:slicer
            end #:flanger
            
          end #:slicer
          
          sleep ((sample_duration sampleRhodes120bpm) * 4.0)
          
        end
        
      end #:song3
      
    end #:ixi
  end #:echo
  
  #sleep 24
  
  live_loop :scream1 do
    
    with_bpm mybpm do
      
      sample sampleClarinetf6, pitch: rrand(-36,-12),
        window_size: (ring 0.003333,0.002,0.001,rrand(0.0001,0.01))[tick],
        beat_stretch: (ring 16,1,2,1,4,1,2,1,8)[tick] * 2,
        pan: rrand(-0.2,0.2), pan_slide: (ring 0.5,1,2,4,8)[tick],
        rate: [-1,1].choose, amp: rrand(1.0,1.25)
      
      sleep (sample_duration sampleClarinetf6) *0.25
      
      sample sampleClarinetf6, pitch: rrand(-36,-20),
        window_size: (ring 0.003333,0.002,0.001,rrand(0.0001,0.01))[tick],
        beat_stretch: (ring 1,2,1,4,1,2,1,8)[tick] * 2,
        rate: [-1,1,-1].choose, amp: 0.25 if one_in(8)
      
      sleep (sample_duration sampleClarinetf6) *0.5
      
      sample sampleClarinetf6, pitch: rrand(-24,-12),
        window_size: (ring 0.003333,0.002,0.001,rrand(0.0001,0.01))[tick],
        beat_stretch: (ring 1,2,1,4,1,2,1,8)[tick] * 2,
        rate: [-1,1,1,-1].choose, amp: 0.25 if one_in(6)
      
      sleep (sample_duration sampleClarinetf6) *0.25
      
      with_fx :slicer, phase: 0.25 do
        sample sampleClarinetf6, pitch: rrand(-36,-16),
          window_size: (ring 0.003333,0.002,0.001,rrand(0.0001,0.01))[tick-1],
          beat_stretch: (ring 4,8,16,32,64)[tick] * 2,
          pan: rrand(-0.5,0.5), pan_slide: (ring 4,8,16,32,64)[tick],
          rate: [1,-1,1].choose, amp: 0.25 if one_in(16)
        
      end #:slicer
      
      sleep (sample_duration sampleClarinetf6) * [1,2,4].choose
      
    end
    
  end #if
  
  #######################################################################
  
  live_loop :scream2 do
    
    if bar > (12+16) and bar < (36+16) then # the break
      
      with_bpm mybpm do
        
        sample sampleClarinetf6, pitch: rrand(-36,-12),
          window_size: (ring 0.003333,0.002,0.001,rrand(0.0001,0.01))[tick],
          beat_stretch: 16,
          pan: rrand(-0.2,0.2), pan_slide: (ring 0.5,1,2,4,8)[tick],
          rate: [-1,1].choose, amp: rrand(1.0,1.25) * 2
        
        sleep (sample_duration sampleClarinetf6) *0.25
        
        sample sampleClarinetf6, pitch: rrand(-36,-20),
          window_size: (ring 0.003333,0.002,0.001,rrand(0.0001,0.01))[tick],
          beat_stretch: 8, rate: [-1,1,-1].choose, amp: 0.5
        
        sleep (sample_duration sampleClarinetf6) *0.5
        
        sample sampleClarinetf6, pitch: rrand(-24,-12),
          window_size: (ring 0.003333,0.002,0.001,rrand(0.0001,0.01))[tick],
          beat_stretch: 4, rate: [-1,1,1,-1].choose, amp: 0.25
        
        sleep (sample_duration sampleClarinetf6) *0.25
        
        with_fx :slicer, phase: 0.25 do
          sample sampleClarinetf6, pitch: rrand(-36,-16),
            window_size: (ring 0.003333,0.002,0.001,rrand(0.0001,0.01))[tick-1],
            beat_stretch: (ring 4,8,16,32,64)[tick] * 2,
            pan: rrand(-0.5,0.5), pan_slide: (ring 4,8,16,32,64)[tick],
            rate: [1,-1,1].choose, amp: 0.25 if one_in(16)
        end #:slicer
      end #bpm
      
    end #if
    
    sleep (sample_duration sampleClarinetf6) * [1,2,4].choose
    
  end #:scream2
  #######################################################################
end #:compressor
#######################################################################

live_loop :hat do
  x = [0.05,0.1].choose # [0.25/2.0,0.25/2.0].choose #offset
  
  if bar < (10+16) or bar > (23+16) then
    
    with_fx :echo, phase: [0.25,0.5].choose,
    mix: rrand(0.5,0.9) do
      with_fx :reverb, mix: 0.2 do
        use_synth :cnoise
        play rrand(90,120), attack: 0.05, release: rrand(0.1,0.12),
          sustain: 0.05,
          amp: rrand(0.10,0.16) *0.6
        
        sleep x #offset
        
        use_synth :pnoise
        play rrand(90,120), attack: 0.05, release: rrand(0.05,0.07),
          sustain: 0.05,
          amp: rrand(0.10,0.14) *0.6
        
      end
    end
  end #if
  
  sleep ((ring 1,0.5,1,0.25)[tick] * 1) - x
  sleep 4 if one_in(64)
  
end #hat
#############################################################

live_loop :kick do
  
  if bar < (8+16) or bar > (22+16) then
    if bar > 4 then
      if bar < 22 or bar % 13 == 0 then
        with_fx :slicer, phase: 0.25 do
          with_fx :ixi_techno, mix: 0.4, res: rrand(0.1,0.3),
          phase: (ring 0.25,48,0.5,24,16,8,4)[tick/4] do
            with_fx :echo, phase: 1.0 / 4.0, reps: 2, mix: 0.666 do
              sample :bd_fat,
                amp: rrand(0.6,0.8) * (ring 0.4,0.5,0,0.4,0.2)[tick/4] * 4,
                pan: rrand(-0.333,0.333), pan_slide: 1.0 / 8.0
              sleep 1.0
            end #fx
          end #fx
        end #fx
      else
        sample :bd_zome,
          amp: rrand(0.8,1.0) * (ring 0.2,0.5,0,0.333,0)[tick/8] * rrand(1.2,1.5),
          rate: -(1.0/[1.0,2.0,3.0,4.0,8.0].choose), window_size: rrand(0.001,0.002),
          beat_stretch: (ring 0.5,1,2,4,8)[tick/4]
        with_fx :reverb, room: 0.5 do
          sample [:drum_bass_hard,:drum_bass_soft].choose, amp: (ring 0,1,1.5,2.0,0)[tick/4]
        end #fx
        sleep 1
      end #if
      sample :bd_klub, amp: rrand(0.8,1.0) *6.4
    end #if
  end #if
  
  sleep 1
  
  sleep 4 if bar > (24+16) and one_in(24)
  
end #kick
#############################################################